### ISSUE TITLE
Jedi locks pyd files preventing recompilation

### ISSUE THREAD

**Issue by segevfiner** [davidhalter/jedi#1211]
If you are developing an extension module for Python and build it in-place (Or use setup.py develop/editable install which also does this) and than your editor requests completions for something involving your built extension module, Jedi will load in the completion process it which will prevent it from being rebuilt on Windows (Extension modules can't be unloaded).

To reproduce:
1. Download [hello_mod.zip](https://github.com/davidhalter/jedi/files/2337441/hello_mod.zip).
2. Build it in-place: `python setup.py build_ext --inplace`.
3. Run `python test.py`. (Requires `jedi`, of course, and `six`).
4. Try to rebuild the module `python setup.py build_ext --inplace`.
5. You will get the following error:
    ```cmd
    C:\hello_mod>python setup.py build_ext --inplace
    running build_ext
    copying build\lib.win-amd64-3.7\_hello.cp37-win_amd64.pyd ->
    error: could not delete '_hello.cp37-win_amd64.pyd': Access is denied
    ```

Originally found here https://github.com/Microsoft/vscode-python/issues/2458

cc @brettcannon

**Comment by davidhalter** [comment]
So what's your advice there, what should Jedi do? The whole point about Jedi loading those modules is to have autocompletion for those modules.

**Comment by segevfiner** [comment]
One way would be to load extension modules in a subprocess, so that the subprocess can be closed in order to free them. Though I'm uncertain what the exact policy of when to open/close such a process should be...

**Comment by davidhalter** [comment]
It is already loaded in a subprocess, however I'm not willing to close that subprocess. Is there no way to free the resources? I think we probably just don't do that... There's probably something wrong in Jedi.

I have a strong feeling that the file is just not closed in `jedi/evaluate/compiled/subprocess/functions.py` (search for `get_module_info`). Could you just try to add a `if module_file is not None: module_file.close()`? That might fix the issue.

**Comment by segevfiner** [comment]
There is already a call to `module_file.close()`:

https://github.com/davidhalter/jedi/blob/84b07a88091c31e55a3a3886412d30a9929ae406/jedi/evaluate/compiled/subprocess/functions.py#L65

And even if you forgot that, it should really just have amounted to a `ResourceWarning`...

If you import an extension module, it remains loaded even if it has no references (You remove it from `sys.modules` and any other reference), with no supported way to unload it (Unloading it directly using OS APIs is unsafe and will leak global resources that the module allocated). Python never calls `FreeLibrary`/`dlclose`, and until Python 3 there wasn't even a callback for a module to free it's resources (`PyModuleDef.m_free`).

**Comment by davidhalter** [comment]
I know that there is this reference, but maybe it's looked at as a package? We are definitely not cleaning up module_files properly (though that is probably not the issue). I don't see why Python should lock extension files if the file handle is closed. But then again I have no idea how dynamic loading exactly works.

It's definitely not the case on Linux (in your example, I cannot reproduce it), so it seems to be a Windows only issue.

**Comment by segevfiner** [comment]
Yes, it's definitely Windows only. When you load an extension module (`LoadLibrary`), Windows doesn't allow to modify or delete the file as long as it's loaded. Linux, on the other hand, does allow that.

And, as I said, the module remains loaded for the lifetime of the process once imported by Python.

**Comment by davidhalter** [comment]
I have no idea what we could do there. Maybe we could copy the module to the temp directory and then load it? But that also feels like a very bad idea.

**Comment by davidhalter** [comment]
Would you be fine with an option of disabling builtin/extension loading in Jedi? Would that solve your issue?

**Comment by pfaion** [comment]
I'm having the same problem. It's really annoying since I have 2 VSCode instances open for development and I have to close them both all the time in order to be able to rebuild the extension modules.

I have to admit though, I don't have an idea of what you could do about it. From a very naive user perspective I would expect jedi to "lock" the .pyd for inspection every time it changes, parsing the information and then releasing it once it's done. I'm sure though that's not so easy if not impossible.

**Comment by segevfiner** [comment]
One way is for Jedi to analyze extension modules in a subprocess so as to not lock them. It would have also helped if VS Code's extension includes a command to restart Jedi but that's outside of this project and expecting each project using Jedi to include such a restart command is hard, sometimes it's not even possible, e.g. IPython.

**Comment by davidhalter** [comment]
The main issue with this is IMO that Python doesn't allow us to free the module again.

The only feasible solution is probably to "restart" Jedi if you don't want to solve the actual problem. For IPython this is obviously also not possible, because IPython imports Jedi (you would have to restart IPython).

However for Jedi used with VScode this could be possible, because Jedi already uses submodules (they are called environments) you can specify a new environment when running Jedi, which automatically **kills** the **old** subprocess. This is basically what you want.

It's not even hard, just create a new environment, work with it and then make sure it is garbage collected if you want to free resources (refcount == 0 should probably be enough). This will probably solve your problem.

**Comment by davidhalter** [comment]
The only drawback of this solution is basically that you have to do the environment management yourself (which is probably what you want, but it can be hard to get this code into VScode).

**Comment by pfaion** [comment]
So is this a "bug" or missing feature in CPython?

I found this post about how hard it is to unload a module in python: https://stackoverflow.com/questions/437589/how-do-i-unload-reload-a-module/487718#487718

**Also just for reference for anyone interested in this and finding this thread:**

I digged a bit and found [PEP 3121](https://www.python.org/dev/peps/pep-3121/) which specified the C interface for module initialization/finalization for Python 3.0+. It references a specific [email conversation](https://mail.python.org/pipermail/python-3000/2006-April/000726.html) involving the BDFL suggesting adding hooks to module finalization e.g. for

> - when the module object is deleted from sys.modules

but apparently there was never time for this. Note this email appears to be from 2001!

Another [issue thread](https://bugs.python.org/issue9072) on the bug tracker deals with this topic and the same guy involved in the previous email responded with:
> In 3.x, chances are slightly higher. In principle, unloading could be supported - but no module actually adds the necessary code, and the necessary code in the import machinery isn't implemented in 3.2 and earlier.
> 
> Supporting unloading will be (and was) a multi-year project. Don't expect any results in the next five years.

Note that this is also from 2010 now...

The [docs for the C-API](https://docs.python.org/3/c-api/init.html#c.Py_FinalizeEx) mention also the lack of an extension module unloading functionality and the general problems that arise with module unloading the way that CPython currently handles them.

So it really does not seems like a simple problem and there seems to be no easy solution. And apparently nobody has time for tackling this since 2001. So sad, I had hoped there would be a simple fix. :disappointed_relieved:

**Comment by davidhalter** [comment]
@pfaion Thanks for the research. It really looks a bit unfortunate.

As I said, I think the workaround with environments is really possible, you just have to find a way to release the current environment.

**Comment by davidhalter** [comment]
I'm closing this one, because I don't really think there's a solution within Jedi. This is mostly a Python issue. If you control Jedi Environments you can just create a different environment and it will work. So Jedi already does its best. 

The only thing that I could basically give you guys is a setting that avoids loading compiled modules completely. That's very possible and that would solve this issue. If you want this, please open a new issue, because it's really not solving this very problem (it's just avoiding it).

**Comment by davidhalter** [comment]
For security reasons, I might have found a solution for you guys anyway. It's on the project branch and will be released in 0.17.0.

Checkout the last comment in #786.

IMO this fixes your issues if the path is not in your sys path. Let me know if that helps (you should be able to work with the project branch easily).
